compiler refactor/eunit & ct fixes

talentdeficit opened the pull request:
changes:

* `extra_src_dirs` are copied from the project to `_build` either relative to the app they are relative to in the project or relative to `_build/profile/extras`. they're compiled in place. copied instead of symlinked so profiles can't leak into one another (and also to keep beams out of the project directories) and compiled in place because `ct` can't handle them being in `ebin`. `rebar_file_utils` calls out to shell to copy files but this loses metadata so anything in an extra dir is going to be recompiled every single time compile is invoked but there's a metadata preserving copy in ec_file that could be used instead. that can be a separate change though

* any directory in `_build` that is used as a compiler output target is added to the code_path/all_deps. i believe this is the right key, but open to correction here. the compiler currently only returns `ok` (or an exception) and changing it to return output dirs (or the application/state object compiled with a field for artifact directories, maybe?) would simplify things (and prevent having to repeatedly recalculate output dirs based on `src_dirs`/`extra_src_dirs`) but that's a refactoring that can wait

* the compiler has a `recursive` option. it's not used anywhere, but it could be controllable with a `{compiler, [{recursive, false}]}` entry in `rebar.config`. omitted for now for pending discussion but i think necessary for having code in ct data dirs, for example

* eunit scopes changes to `eunit_compile_opts`, `eunit_first_files` and files and directories in the test set properly. i'm unsure how the `rebar.config` in an app directory (as opposed to the project root) is supposed to apply to the app so there's probably some missing testing around this but that too can be fixed in a separate change. the `eunit` provider also translates paths used in test sets so doing `rebar3 eunit --dir=test` will now ensure `test` is compiled and call `eunit:test([{dir, "_build/test/extras/test"}])`. `rebar3 eunit --dir=apps/foo/test` will call `eunit:test([{dir, "_build/test/lib/foo/test"}])`

talentdeficit commented:
oops, just noticed this omits one last commit that fixes `clean` in erlc_compiler but it's on my other laptop. i'll add it tonite

talentdeficit commented:
r15 is the bane of my existence. i'll figure that out tonite too

tsloughter commented:
Looks good. @talentdeficit as for the eunit paths in a rebar.config in a subapp, the subapp rebar.config should always be treated like it could be standalone. So it is relative to that app.

ferd commented on src/rebar_erlc_compiler.erl:
```diff
@@ -26,9 +26,9 @@
--export([compile/1,
-         compile/3,
-         compile/4,
+-export([compile/1, compile/2, compile/3,
```
Isn't getting rid of `compile/4` possibly messing with plugins?

talentdeficit replied:
i don't think so. i added it a month ago ([here][https://github.com/rebar/rebar3/commit/9d82215296205d2d30c152c7d7c6fcb8e4a92443]) for the ct provider but it no longer needs it with the changes to the compiler

ferd commented on src/rebar_erlc_compiler.erl:
```diff
@@ -81,111 +89,189 @@
+%% @equiv compile(AppInfo, []).
+
+-spec compile(rebar_app_info:t()) -> ok.
+compile(AppInfo) when is_tuple(AppInfo), element(1, AppInfo) == app_info_t ->
```
can get rid of the whole guard with `is_record(AppInfo, app_info_t)`, or at the very least just drop `is_tuple/1` given `element/2` encompasses it.

talentdeficit replied:
i'll drop the `is_tuple/1` but currently the state_t and app_info_t records aren't exported

ferd commented on src/rebar_erlc_compiler.erl:
```diff
@@ -81,111 +89,189 @@
+%% @doc compile an individual application.
+
+-spec compile(rebar_app_info:t(), compile_opts()) -> ok.
+compile(AppInfo, CompileOpts) when is_tuple(AppInfo), element(1, AppInfo) == app_info_t ->
```
Same guard comment here.

ferd commented on src/rebar_erlc_compiler.erl:
```diff
@@ -81,111 +89,189 @@
+%% @hidden
+
+-spec compile(rebar_dict() | rebar_state:t(), file:name(), file:name(), compile_opts()) -> ok.
+compile(State, BaseDir, OutDir, CompileOpts) when is_tuple(State), element(1, State) == state_t ->
```
Same guard comment.

ferd commented on src/rebar_erlc_compiler.erl:
```diff
@@ -81,111 +89,189 @@
+    end,
+    lists:foreach(F, lists:map(fun(SrcDir) -> filename:join(Dir, SrcDir) end, ExtraDirs)).
+
+%% @hidden
```
I think this could go in `@doc` without a problem. We can make the intent specific.

ferd commented on src/rebar_prv_eunit.erl:
```diff
@@ -73,11 +74,12 @@ do_tests(State) ->
-    ?DEBUG("eunit_opts  ~p", [EUnitOpts]),
-    Result = eunit:test(Tests, EUnitOpts),
+    ?INFO("eunit_tests ~p", [T]),
+    ?INFO("eunit_opts  ~p", [EUnitOpts]),
```
Is this intentionally left this way, or that was helping filter out the info you needed from the flood of `DEBUG` rebar3 generates?

talentdeficit replied:
yeah, this is just an oversight on my part. i'll change this

ferd commented:
Got a bunch of comments, but they're all about fairly minor stuff but one. Nice changeset.

talentdeficit commented:
i rebased this on master, squashed all the non-compiling commits and made a minor non-functional cleanup change to the eunit provider

pending approval of dropping `compile/4` (which i don't think has been around long enough to worry about keeping) this is good to merge i think

ferd commented:
Yes, I'm fine with dropping compile/4. We can merge whenever. @tsloughter you'Re okay with merging this?
